qemu/tests/qemu_guest_agent_update.py

```diff
@@ -0,0 +1,150 @@
+            while time.time() < end_time:
+                time.sleep(2)
+                virtio_win_letter = utils_misc.get_win_disk_vol(session,
+                                                                vol_virtio_key)
+                if virtio_win_letter:
+                    break
+            if not virtio_win_letter:
+                test.fail("Couldn't get virtio-win volume letter.")
+
+        def _get_pkg_download_cmd():
+            """
+            Only for url method to get the previous qemu-ga version and get the
+            download cmd for this pkg.
+
+            """
+            qga_html = params["qga_html"]
+            qga_url = params["qga_url"]
+            qga_html_download_cmd = "wget %s -O %s" % (qga_url, qga_html)
+            process.system(qga_html_download_cmd,
+                           float(params.get("login_timeout", 360)))
+
+            with open(qga_html, "r") as f:
+                lines = f.readlines()
+            list_qga = []
+            for line in lines:
+                if "qemu-ga-win" in line:
+                    list_qga.append(line)
+            tgt_line = list_qga[-2]
+            # qemu-ga-win-7.5.0-2.el7ev
+            pattern = re.compile(r"qemu-ga-win-\d\.\d\.\d-\d.el\dev")
+            qga_pre_pkg = pattern.findall(tgt_line)[0]
+            logging.info("The previous qemu-ga version is %s." % qga_pre_pkg)
+
+            # https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/
+            #   archive-qemu-ga/qemu-ga-win-7.5.0-2.el7ev/
+            qga_url_pre = r"%s/%s/%s" % (qga_url, qga_pre_pkg,
+                                         self.qemu_ga_pkg)
+            qga_host_path = params["gagent_host_path"]
+            params["gagent_download_cmd"] = "wget %s -O %s" % (qga_url_pre,
+                                                               qga_host_path)
+
+        def _qga_install():
+            """
+            Install qemu-ga pkg.
+
+            """
+            qga_pkg_path = self.get_qga_pkg_path(
+                self.qemu_ga_pkg, test, session, params, vm)
+            self.gagent_install_cmd = params.get("gagent_install_"
+                                                 "cmd") % qga_pkg_path
+            self.gagent_install(session, vm)
+
+        error_context.context("Boot up vm.", logging.info)
+        params["start_vm"] = "yes"
+        latest_qga_download_cmd = params["gagent_download_cmd"]
+        env_process.preprocess_vm(test, params, env, params["main_vm"])
+        vm = self.env.get_vm(params["main_vm"])
+        vm.verify_alive()
+        session = self._get_session(params, vm)
+
+        error_context.context("Install the previous qemu-ga in guest.",
+                              logging.info)
+        if self._check_ga_pkg(session, params["gagent_pkg_check_cmd"]):
+            logging.info("Uninstall the one which is installed.")
+            self.gagent_uninstall(session, vm)
+
+        if self.gagent_src_type == "virtio-win":
```

Comment:
I don't find anywhere utilize it, is this "virtio-win" src_type actually needed?

Reply:
yes,in internel test,we get qemu-ga-win pkg via virtio-win cdrom,in upstream they get qemu-ga-win via fedora url.